web/pdf_rendering_queue.js

```diff
@@ -119,12 +120,21 @@ class PDFRenderingQueue {
     if (numVisible === 0) {
       return null;
     }
+
     for (let i = 0; i < numVisible; i++) {
       const view = visibleViews[i].view;
       if (!this.isViewFinished(view)) {
         return view;
       }
     }
+
+    for (let i = 0; i < numVisible; i++) {
+      const { detailView } = visibleViews[i].view;
+      if (detailView && !this.isViewFinished(detailView)) {
+        return detailView;
+      }
+    }
+
```

Comment:
*Another observation/question that I have here is about the order of rendering.*

Would it make sense to attempt to render both a page *and* its `detailView` before moving onto the other visible pages, such that the most visible page will always render "fully" without having to wait for (potentially) a lot of other pages to finish rendering first?

Basically something like the following, where we combine the loops.
However, please note that I'm *not* saying that we must do this but I rather just want us to have considered/discussed this!
```suggestion
    for (let i = 0; i < numVisible; i++) {
      const view = visibleViews[i].view;
      if (!this.isViewFinished(view)) {
        return view;
      }
      const { detailView } = view;
      if (detailView && !this.isViewFinished(detailView)) {
        return detailView;
      }
    }

```

Reply:
The answer is _maybe_.

The main use case I am thinking about for this patch is maps, because they are large documents where you usually want to zoom in a lot to the particular point that you are interested in.

The assumption is that the normal usage flow is:
- you zoom in a lot to roughly the point you are interested in
- you scroll a bit to adjust what's visible, so that you can actually see what you want
- you might scroll more in any direction, for example to follow a street or to see what is close to the point you were looking at

It's thus important that when you scroll around the page there is _something_ visible, and so this patch prioritizes rendering the "background view" over the "detail view" (a second reason is that it makes it easier to integrate https://github.com/mozilla/pdf.js/pull/19043 because it means that when we try to render the detail view we already have the bounding boxes, but there are also other ways to do it). This however matters _within a page_: it's probably unlikely that you zoom in to the wrong page, and then you need to scroll to another one. So your suggestion would be reasonable and not in conflict with the use case.

However, there is a problem: the visible pages are rendered first-to-last, meaning that if you have 10px of page 1 visible and 500px of page 2, it will first render page 1 and then page 2. Rendering page 1, then the detail of page 1, and then page 2 increase the total time for seeing _something meaningful_ on screen. So we would first need to change `getHighestPriority` to prioritize pages by how much they are visible, and _then_ we can do something like that you are suggesting.

> wait for (potentially) a lot of other pages

Note that it's _very likely_ that there are at most two pages on screen when the detail view is relevant: the pages must be so large that are above `maxCanvasPixels`, which makes it difficult to fit more than part of a page in the top half of the screen and part of it in the bottom half.

The only realistic case I can think of is when there are two pages side-by-side, so that we have four pages in total (one per "corner").

Reply:
> So we would first need to change `getHighestPriority` to prioritize pages by how much they are visible, and _then_ we can do something like that you are suggesting.

Isn't that already the case though, given the `sortByVisibility` parameter?
https://github.com/mozilla/pdf.js/blob/2df5d8f3ef65a16ef0eb13668e64d74891b419d0/web/ui_utils.js#L423-L425

Reply:
*As mentioned above, my intention here is only that we make an informed decision (whatever it may be) by considering the various cases at hand.*

> I have a slight preference for the second, because some PDFs can take very long (30-60 seconds per page) and so the quicker we show something the better it is, but ultimately I don't feel strongly about it.

There's only additional complication if there's multiple pages visible *and* they contain very large image resources.

We have API caching of operatorLists and image (and other) resources, however for complex pages those caches will be cleared (with a delay) after rendering completes. So by delaying rendering of the `detailView` until other pages have finished there's a risk that the caches have been cleared, and that we'll need to re-parse the page on the worker-thread for the `detailView` rendering.

See e.g. the following API code (and also the `_maybeCleanupAfterRender` property):
https://github.com/mozilla/pdf.js/blob/2df5d8f3ef65a16ef0eb13668e64d74891b419d0/src/display/api.js#L1780-L1787